# **语法**

JS 语法定义了词法规则（如运算符和关键词等）是如何构成可运行的程序代码的。

换句话说，只看词法不看语法会遗漏掉很多重要的细节。

本章介绍的是语法，虽然和开发人员直接打交道的是词法。

# 5.1 语句和表达式

开发人员常常将"语句"（statement）和“表达式”（expression）混为一谈。但是他们在 JS 中存在重要的差别。

区别：

JS，**语句相当于一个句子**，而**表达式就相当于句子中的短语**，运算符相当于句子中的标点符号和连接词。

JS 中表达式可以返回一个结果值。

```js
var a = 3 * 6;
var b = a;
b;
```
这里 3*6 是一个表达式(结果为18)。

第二行的 a 也只是一个表达式，第三行的 b 也是表达式。表达式 a 和 b 的结果都是18.

这三行代码都是包含表达式的**语句**。

var a = 3 * 6; ，var b = a;称为“声明语句”。因为它们声明了变量。

a = 3 * 6、b = a叫做赋值表达式。

第三行代码中只有一个表达式 b，同时它也是一个语句。虽然没什么意义，通常被称为“表达式语句”


## **5.1.1 语句的结果值**

**语句都有一个结果值**（undefined 也算）。 

获得结果值最直接的方法是在浏览器控制台中输入语句，默认情况下控制台会显示所执行的最后一条语句的结果值。

以赋值表达式 b = a 为例，其结果值是赋给 b 的值(18),但规范定义 var 的结果值是 undefined。

如果在控制台中输入 var a = 42;会得到结果值 undefined，而非42。

但是在我们代码中是没办法获得结果值的。说先弄清楚为什么要获得语句的结果值。

闲来看看其他语句的结果值。比如代码块{..}的结果值是它最后一个语句/表达式的结果。

例如：
```js
var b;

if (true) {
    b = 4 + 38;
}
```
在控制台会显示42，即最后一个语句/表达式 b = 4 + 38的结果值。

换句话说，代码块的结果就会如同一个**隐式的返回**，即返回最后一个语句的结果值。

但下面这样的代码无法运行：
```js
var a,b;

a = if(true) {
    b = 4 + 38;
};
```
因为语法不允许我们获得语句的结果值并将其赋值给例外一个变量。

获得语句的结果值的方法（ES5之前），使用 万恶的eval()来获取语句的结果值。（千万不要这么做）
```js
var a,b;
a = eval("if(true){b=4+38;}");
a;
```

ES7(我去，都7了)规范有一项"do 表达式"提案：

类似下面这样
```js
var a, b;
a = do {
    if(true) {
        b = 4 + 38;
    }
};
a;      //42
```

不过现在浏览器肯定没支持.

其目的是将语句当作表达式来处理了（为了语句中可以包含其他语句），从而不需要将语句封装为函数在调用 return 来返回值。

## **5.1.2 表达式的副作用**

大部分表达式没有副作用。例如：

```js
var a = 2;
var b = a + 3;
```

表达式 a + 3本身没有副作用。它的结果值为5，通过 b = a + 3赋值给变量 b。

最常见的副作用的表达式是函数调用。
```js
function foo() {
    a = a + 1;
}

var a = 1;

foo();      // 结果值为：undefined。副作用：a 的值被改变（因为 a 被重新赋值了）
a;          //2
```

递增运算符++ 和递减运算符 -- 都是医院运算符，它们既可以在前面，又可以用在右面
```js
var a = 42;
var b = a++;

a;  // 43
b;  // 42
```
++ 在前面时， 如 ++a，它的副作用（将 a 递增）产生在表达式返回结果值之前。而 a++ 的副作用则产生在返回结果之后。

要想 b 的值是43，那么应该使用逗号运算符将多个独立的表达式串联成一个语句：
```js
var a = 42, b;
b = (a++, a);
a;  // 43
b;  // 43
```
由于运算符优先级的关系，a 需要放在( )里面。

a++，a 中的第二个表达式a 在 a++之后执行，结果为43，并被赋值给 b。

再如 delete 运算符。

delete 用来删除对象中的属性和数组中的单元。它通常以单独一个语句的形式出现：
```js
var obj = {
    a: 42
};

obj.a;          // 42
delete obj.a;   //true
obj.a           //undefined
```
如果操作成功，delete 返回 true，否则返回 false。

其副作用是属性被从对象中删除。

另外一个有趣的例子是 = 赋值运算符。
```js
var a;
a = 42;
a;      //42
```
a = 42 中的=运算符看起来没有副作用，实际上它的结果值是42，它的副作用将42赋值给 a。

多个赋值语句串联时（链式赋值），赋值表达式（和语句）的结果值就能派上用场，比如：
```js
var a, b, c;
a = b = c = 42;
a;  //42
b;  //42
c;  //42
```

另一个需要注意的问题是：
```js
function vowels(str) {
    var matches;
    if(str) {
        matches = str.match(/[aeiou]/g);
        if(matches) {
            return matches;
        }
    }
}

vowels("Hello World");  //["e","o","o"]
```

上面的代码没问题。但我们可以利用赋值语句的副作用将两个 if 语句合二为一：
```js
function vowels(str) {
    var matches;
    if(str && (matches = str.match(/[aeiou]/g))) {
        return matches;
    }
}
vowels("Hello World");  //["e","o","o"]
```

### **5.1.3 上下文规则**

在 JavaScript 语法规则中，有时候同样的语法在不同的情况下会有不同的解释。

### 1. 大括号

下面两中情况下会用到大括号{..}。

**（1）对象常量**

用大括号定义对象常量。

var a = {
    foo: bar()
};

{ .. }被赋值给 a，因而它是一个对象常量。

### 2. 代码块

还有一个坑常被提到（涉及强制类型转换）
```JS
[] + {};    //"[object Object]"
{} + [];    // 0
```

{}出现在 + 运算符后面，这是它被当作一个值（空对象处理）。[]会被强制类型转换为" "，所以{}会被强制类型转换为"[object Object]"

而
```js
{} + [];

//实质就等于
{}      //类似 function(){} 这种，后面不用加分号
+[];    //  O
```

代码块结尾不需要分号(函数表达式除外)，所以不存在语法上的问题，最后+[]将显示强制类型转换。

### 3. 对象解构

从 ES6 开始， {..}也可以用于“解构赋值”，特别是对象的结构。

```js
function getData() {
    //..
    return {
        a : 42,
        b: "foo"
    };
}

var { a,b } = getData();

console.log(a, b);   //42 "foo"
```

{a, b} = .. 就是 ES6中的解构赋值，相当于下面的代码：
```js
var res = getData();
var a = res.a;
var b = res.b;
```

{a, b}实际上是{a:a, b:b}的简化版本。

{ .. }还可以用作函数命名参数的对象结构，方便隐式地用对象属性赋值：
```js
function foo({a, b, c}) {
    // 不再需要这样：
    // var a = obj.a, b = obj.b, c = objc.c
    console.log(a,b,c);
}

foo(
    {
        c: [1,2,3],
        a: 42,
        c: "foo"
    }
)   // 42 "foo" [1,2,3]
```

### 4. else if和可选代码块

很多人误以为 JS 中有 else if。事实上并没有，但 if 和 else 只包含单挑语句的时候可以省略代码块{ }。

比如：
```js
if (a) doSomething(a);
```

但建议加上{}:
```js
if (a) {doSomething(a);}
```

## **5.2.3 关联**

&& 和 || 运算符先于 ?:执行，那么如果多个相同优先级的运算符同时出现时，又改如何处理呢（比如多个&&， `a && b && c`）?

一般来说，运算符的关联不是从左到右就是从右到左，这取决于是从左开始还是右开始。

关联和执行顺序不是一回事。

但它为什么又和执行顺序有关？原因是表达式可能会产生副作用，比如函数调用：
```js
var a = foo() && bar();
```

这里 foo() 首先执行，它的返回结果决定了 bar()是否执行。

所以如果 bar()在 foo() 之前执行，整个结果会完全不同。

这里遵循从左到右的顺序（JS 默认执行顺序），与&&的关联无关。

因为上例子中只有一个`&&`运算符，所以不涉及组合和关联。

```js
var a = true;
var b = true;
var c = false;

(a && b) && c;

a && (b && c);
```

&& 和 || 运算符是左关联的。

不过`？:`三元运算符则是右关联（当出现2对`？:`时，右边的先）。

比如：
```js
a ? b : c ? d : e;
```
实际上关联顺序是：
```js
a ? b : (c ? d : e);
```
`？:`三元运算符则是右关联，它的组合方式会影响返回的结果。

另一个右关联（组合）的例子是`=`运算符。

```js
var a, b, c;
a = b = c = 42;
```

它首先执行 c = 42，然后 b = .. ，最后是 a = ..。因为是右关联，所以他实际处理的是：`a = (b = (c = 42))`

# 5.3 自动分号

JS 会自动为代码行补上缺失的分号，即自动分号插入 （ASI）

但是为了规范，尽量不依赖这种方式，该有的分号必须有。


ASI 就是一个纠错机制，为了提高容错率来实现代码的可解析。

# 5.4 错误

在编译阶段发现的代码错误叫“早期错误”。语法错误是早期错误的一种。

1. 语法规定赋值对象必须是一个标识符：
```js
var a;
42 = a; //错误！！
```

2. 严格模式中，函数的参数不能重名：
```js
function foo(a,b,c) {}              //没问题
function bar(a,b,a) {"use strict";} //错误
```

3. 对象常量不能包含多个同名属性：
```js
(function() {
    "use strict";
    var a = {
        b: 42,
        b: 43
    };  //错误
})();
```
从语义角度来说，这些错误并非词法错误，而是语法错误，因为它们在词法上是正确的。

## 提前使用变量

ES6规范定义了一个新概念，叫作 TDZ（Temporal Dead Zone，暂时性死区）。

TDZ 指的是由于代码中的变量还没有初始化而不能被引用的情况。

最直观的例子是 ES6规范的 let 块作用域：
```js
{
    a = 2;  // ReferencError!
    let a;
}
```
a = 2 试图在 let a 初始化 a 之前使用该变量（其作用域在{..}内），这里就是 a 的 TDZ，会产生错误。

有意思的是，对未声明变量使用 typeof 不会产生错误，但在 TDZ 中却会报错：
```js
{
    typeof a;   // undefined
    typeof b;   //ReferenceError! (TDZ)
    let b;
}
```


# 5.5 函数参数

ES6中，可以为函数参数设置默认值，当现实被省略或者值为 undefined 时，则取该函数的默认值：
```js
function foo(a = 42, b = a + 1) {
    console.log(a, b);
}
foo();  // 42 43
foo(undefined); //  42 43
foo(5); // 5 6
foo(void 0, 7); // 42 7
foo(null);      // null 1
```

对 ES6中的参数默认值而言，参数被**省略**或被赋值为 `undefined` 效果都是一样的，都是取该参数的默认值。

然而某些情况下，它们之间还是有区别的：
```js
function foo(a = 42, b = a+1) {
    console.log(
        arguments.length, a, b,
        arguments[0],arguments[1]
    )
}

foo();  //0 42 43 undefined undefined
foo(10);    //1 10 11 10 undefined
foo(10, undefined); //2 10 11 10 undefined
foo(10, null);      //2 10 null 10 null
```

虽然参数 a 和 b 都有默认值，但是函数不带参数时，arguments 数组为空。

相反，如果函数传递 undefined 值，则 arguments 数组中会出现一个值为 undefined 的单元，而不是默认值。

ES6参数默认值会导致 arguments 数组和对应的命名参数之间出现偏差。

ES5也出出现这种情况：
```js
function foo(a) {
    a = 42;
    console.log(arguments[0]);
}
foo(2); //42 (linked)
foo();  //undefined (not linked) 
```

但是，严格模式中并没有创建关联这一说：
```js
function foo(a) {
    "use strict";
    a = 42;
    console.log(arguments[0]);
}

foo(2); //2 (not linked)
foo();  // undefined(not linkded)
```

因此不要在开发中依赖这种关联机制。实际上他是 JS 语言引擎底层实现的一种抽象泄漏，比不是语言本身的特性。

arguments 数组已经被废止了。但不过它并非一无是处。

因为在 ES6 之前，获得函数所有参数的唯一途径就是 arguments 数组。此外，即使将命名参数和 arguments 数组混用也不会出错，只需遵守一个原则，即不要同时访问命名参数和其对应的 arguments 数组单元。

```js
function foo(a) {
    console.log(a + arguments[1]);  //安全！
}

foo(10, 32);    //42
```

# 5.6 try..finally

try..catch可以和 finally 配对使用，必要时两者可以同时出现。

finally 中的代码总是会在 try 之后执行，如果有 catch 的话，则再 catch 之后执行。

可以将 finally 中的代码看做一个回调函数，即无论出现什么情况最后一定会被调用。

try 中有 return 语句会出现什么情况，return 会返回一个值，那么调用该函数并得到返回值的代码是 finally 之前还是之后执行？

```js
function foo() {
    try {
        return 42;
    }
    finally {
        console.log("hello");
    }
    console.log("never runs");
}

console.log(foo());
// hello
// 42
```
这里 return 42 先执行，并将 foo() 函数的返回值设置为42.然后 try 执行完毕，接着执行 finally。最后 foo() 函数执行完毕，console.log(..)显示返回值。

try 中的 throw 也是如此：
```js
function foo() {
    try {
        throw 42;
    }
    finally {
        console.log("hello");
    }
    console.log("never runs");
}

console.log(foo());
// hello
// Uncaught Exception: 42
```

如果 finally 中抛出异常，函数就会在此终止，如果此前 try 中已经有return 设置了返回值，则该值会被丢弃：
```js
function foo() {
    try {
        return 42;
    }
    finally {
        throw "Oops!";
    }
    console.log("never runs");
}

console.log(foo());
//Uncaught Exception: Oops!
```

continue 和 break 等控制语句也是如此：
```js
for (var i=0; i < 10; i++) {
    try {
        continue;
    }
    finally {
        console.log(i);
    }
}

// 0 1 2 3 4 5 6 7 8 9
```

finally 中 return 会覆盖 try 和 catch 中 return 的返回值：
```js
function foo() {
    try {
        return 42;
    }
    finally {
        // 没有返回语句，所以没有覆盖
    }
}

function bar() {
    try {
        return 42;
    }
    finally {
        //覆盖前面的 return 42
        return;
    }
}

function baz() {
    try {
        return 42;
    }
    finally {
        //覆盖前面的 return 42
        return "hello";
    }
}

foo();  //42
bar();  //undefined
baz();  //hello
```

# 5.7 switch

switch 就是 if..else if..else..的简化版
```JS
switch(a) {
    case 2:
        //执行一些代码
        break;
    case 42:
        //执行一些代码
        break;
    default:
        //执行缺省代码
}
```

这里 a 与 case 表达式注意进行比较。如果匹配就执行 case 中的代码，知道 break 后者 switch 代码块结束。

但有一个不为人知的陷阱。

就是 a 与 case 表达式的匹配算法（两者比较）和严格相等 === 相同。

```js
var a = "42";
switch(true) {
    case a == 10:
        console.log("10 or '10'");
        break;
    case a == 42:
        console.log("42 or '42'");
        break;
    default:
        //执行缺省代码
}
```
// 42 or '42'

因为 a == 42 为 true，所以与 switch 的条件匹配，所以执行。

尽管可以使用==，但 switch 中 true 和 true 之间仍然是严格相等比较。

但如果a 与 case 表达式两者只是宽松相等，而不是严格相等，条件会无法匹配。比如
```js
var a = "hello world";
var b = 10;
switch(true) {
    case (a || b == 10):
        //永远执行不到这里
        break;
    default:
     console.log("oops");
}
```
// oops

因为(a || b == 10)的结果是“hello world”而非 true，所以严格相等比较不成立。

此时可以通过强制表达式返回 true 或 false，如 `case !!(a || b == 10):`

这样能够匹配 true。